substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1163,6 +1159,81 @@ impl<T: Config> Pallet<T> {
 	) -> Exposure<T::AccountId, BalanceOf<T>> {
 		EraInfo::<T>::get_full_exposure(era, account)
 	}
+
+	pub(super) fn do_migrate_currency(stash: &T::AccountId) -> DispatchResult {
+		if Self::is_virtual_staker(stash) {
+			return Self::do_migrate_virtual_staker(stash);
+		}
+
+		let ledger = Self::ledger(Stash(stash.clone()))?;
+		let staked: BalanceOf<T> = T::OldCurrency::balance_locked(STAKING_ID, stash).into();
+		ensure!(!staked.is_zero(), Error::<T>::AlreadyMigrated);
+		ensure!(ledger.total == staked, Error::<T>::BadState);
+
+		// remove old staking lock
+		T::OldCurrency::remove_lock(STAKING_ID, &stash);
+
+		// check if we can hold all stake.
+		let max_hold = asset::free_to_stake::<T>(&stash);
+		let force_withdraw = if max_hold >= staked {
+			// this means we can hold all stake. yay!
+			asset::update_stake::<T>(&stash, staked)?;
+			Zero::zero()
+		} else {
+			// if we are here, it means we cannot hold all user stake. We will do a force withdraw
+			// from ledger, but that's okay since anyways user do not have funds for it.
+			let force_withdraw = staked.saturating_sub(max_hold);
+
+			// we ignore if active is 0. It implies the locked amount is not actively staked. The
+			// account can still get away from potential slash but we can't do much better here.
+			StakingLedger {
+				total: max_hold,
+				active: ledger.active.saturating_sub(force_withdraw),
+				// we are not changing the stash, so we can keep the stash.
+				..ledger
+			}
+			.update()?;
+			force_withdraw
+		};
+
+		// Get rid of the extra consumer we used to have with OldCurrency.
+		frame_system::Pallet::<T>::dec_consumers(&stash);
+
+		Self::deposit_event(Event::<T>::CurrencyMigrated { stash: stash.clone(), force_withdraw });
```

Comment:
Is this event suppossed to notify of how much funds were migrated?
Because by looking at the if-else above it looks like it's more like funds_still_locked_in_old_currency.

Reply:
Not sure if I understand `funds_still_locked_in_old_currency `? Once we get here, there should be nothing locked with old currency. 

The event is meant to notify clients that the given account has migrated to new currency along with the information if their ledger has reduced by some balance because of this operation.

Reply:
I understood that the "else" condition is when someone has more funds locked for staking than in their free balance thus we cannot "hold" as much as is needed so we go just for as much as possible.
Regarding `funds_still_locked_in_old_currency`, I called it so because in `if` zero value is returned and in `else`  it's the difference between staked and free balance and since the free balance will be held then the result is what's left to unlock and move to hold.
Is my reasoning correct?